### PR TITLE
fix: handle timeouts during tls negotiation for `strict` encryption

### PR DESCRIPTION
**Pull request by MichaelSun90**
Network socket not correctly cleaned up when connection timeout fires during TDS8.0 TLS negotiation #1558

### CODE REVIEW

**Comment by MichaelSun90** [comment]
@arthurschreiber remove the TDS 7_1 checks against SQL server 2022 since 7_1 suppose not work with SQL server 2022 and it did failing the encrypt tests. We did no run into this because the added encrypt tests are not actually running due to a error within our test. If we still want to keep the 7_1 test run against SQL server 2022 for certain reason, we can also add some logic to skip the tests like that we did in 'should reset Connection'.

File: src/connection.ts [hunk 1]
```diff
@@ -2008,7 +2009,11 @@ class Connection extends EventEmitter {
     this.transitionTo(this.STATE.SENT_PRELOGIN);
   }
 
-  wrapWithTls(socket: net.Socket): Promise<tls.TLSSocket> {
+  wrapWithTls(socket: net.Socket, signal: AbortSignal): Promise<tls.TLSSocket> {
+    if (signal.aborted) {
+      throw new AbortError();
+    }
```
Comment (arthurschreiber) [inline src/connection.ts:2015]:
```
signal.throwIfAborted();
```

File: src/connection.ts [hunk 2]
```diff
@@ -2023,12 +2028,41 @@ class Connection extends EventEmitter {
...
-        encryptsocket.removeListener('error', reject);
+      const encryptsocket = tls.connect(encryptOptions);
+
+      const onAbort = () => {
+        encryptsocket.removeListener('error', onError);
+        encryptsocket.removeListener('connect', onConnect);
+
+        encryptsocket.destroy();
+
+        reject(new AbortError());
```
Comment (arthurschreiber) [inline src/connection.ts:2039]:
```
reject(signal.reason);
```

File: src/connection.ts [hunk 3]
```diff
@@ -41,6 +41,7 @@ import { createNTLMRequest } from './ntlm';
 import { ColumnEncryptionAzureKeyVaultProvider } from './always-encrypted/keystore-provider-azure-key-vault';
 
 import { AbortController, AbortSignal } from 'node-abort-controller';
+import AbortError from './errors/abort-error';
```
Comment (arthurschreiber) [inline src/connection.ts:44]: You might be able to remove this import if you remove the cases where you construct `AbortError` instances.

**Comment by MichaelSun90** [comment]
Figured out, that signal.reason can be undefined which causing an additional error if we do tostring and compare it with the error that has been caught see if it include the abort title. However, just err === signal.reason still does not handles the case that :
signal.reason :
```
Error: AbortError
    at AbortController.abort (C:\Users\msun\Documents\GitHub\tedious\node_modules\node-abort-controller\index.js:56:31)
    at Timeout._onTimeout (C:\Users\msun\Documents\GitHub\tedious\src\/connection.ts:2116:18)
    at listOnTimeout (node:internal/timers:569:17)
    at processTimers (node:internal/timers:512:7)
```
err is:
```
AbortError: The operation was aborted
    at EventEmitter.onAbort (C:\Users\msun\Documents\GitHub\tedious\src\/connector.ts:108:18)
    at EventEmitter.emit (node:events:514:28)
    at AbortSignal.dispatchEvent (C:\Users\msun\Documents\GitHub\tedious\node_modules\node-abort-controller\index.js:28:23)
    at AbortController.abort (C:\Users\msun\Documents\GitHub\tedious\node_modules\node-abort-controller\index.js:58:17)
    at Timeout._onTimeout (C:\Users\msun\Documents\GitHub\tedious\src\/connection.ts:2116:18)
    at listOnTimeout (node:internal/timers:569:17)
    at processTimers (node:internal/timers:512:7) {
  code: 'ABORT_ERR'
}
```

```
if (err === signal.reason || err.name === 'AbortError'){
    return;
}
```
 ```
if(signal.reason && signal.reason.tostring().include('AbortError')) {
      return;
 }
```
both of these will handle the errors correctly, but not exactly a clean solution in my opinion.
Not sure if we should go with either one or @arthurschreiber  you have suggestions for handling this.

**Comment by arthurschreiber** [comment]
@MichaelSun90 I changed the code to check for `signal.aborted` instead, and I think this now has the correct behaviour.

If the `signal` (which is coming from the `connectTimeout`) was aborted, that means that the connect timeout has fired and we need to ignore that error (because a different error will be generated by the connect timeout code.

Anyway, hopefully this can be cleaned up via the changes in https://github.com/tediousjs/tedious/pull/1562.

**Comment by MichaelSun90** [comment]
> @MichaelSun90 I changed the code to check for `signal.aborted` instead, and I think this now has the correct behaviour.
> 
> If the `signal` (which is coming from the `connectTimeout`) was aborted, that means that the connect timeout has fired and we need to ignore that error (because a different error will be generated by the connect timeout code.
> 
> Anyway, hopefully this can be cleaned up via the changes in #1562.

Thanks for help on this change, and release it ! :bowing_man:

**Pull request merged**